README
+---------------------------------------------------------------------------
|   PHP-IRC v2.2.1 Service Release
|   ========================================================
|   by Manick
|   (c) 2001-2006 by http://www.phpbots.org/
|   Contact: manick@manekian.com
|   irc: #manekian@irc.rizon.net
|   ========================================
|   Special Contributions were made by:
|   cortex
+---------------------------------------------------------------------------
|   > Documentation
+---------------------------------------------------------------------------
|   > This program is free software; you can redistribute it and/or
|   > modify it under the terms of the GNU General Public License
|   > as published by the Free Software Foundation; either version 2
|   > of the License, or (at your option) any later version.
|   >
|   > This program is distributed in the hope that it will be useful,
|   > but WITHOUT ANY WARRANTY; without even the implied warranty of
|   > MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
|   > GNU General Public License for more details.
|   >
|   > You should have received a copy of the GNU General Public License
|   > along with this program; if not, write to the Free Software
|   > Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+---------------------------------------------------------------------------
|   Changes
|   =======-------
|   > If you wish to suggest or submit an update/change to the source
|   > code, email me at manick@manekian.com with the change, and I
|   > will look to adding it in as soon as I can.
+---------------------------------------------------------------------------

NOTE! Please enable word-wrap to view this file.
NOTE! PLEASE READ SECTION '3. Installation - A few things to consider' BEFORE USING.

Please visit our new mods/forums website here: http://www.phpbots.org/

Table of Contents
-----------------

1. 	Introduction and Release Notes
1-a.	What's new in 2.2.1!
1-b.	Features
2. 	Included
3. 	Installation - A few things to consider
3-a.	Quick Install
3-b. 	Installing PHP 5
3-c.	Using alternate PHP.INI (PHP Configuration) file
3-d. 	Installing PHP-IRC
4. 	Configuration
5. 	Running
6. 	Remote Administration (via DCC and Private Message)
7. 	File Transfer
7-a. 	File Transfer behind a firewall or NAT
8. 	Modules and User-defined functions
8-a. 	Submitting modules for others to download
9. 	Querying outside servers (alternative to fopen, fsockopen, etc) READ THIS if you need this functionality!
9-a.	Basic Queries, using the addQuery() function
9-b.	Intro to Connection Class, Advanced Queries
10. 	Custom DCC Chat Sessions
10-a. 	Custom DCC Chat Example: Simple File Server (SFS)
11. 	Database support
11-a.	Mysql Database
11-b.	Postgre Database
11-c.	Serverless, mIRC compatible ini-file based Database
12. 	Timers
13. 	Multiple servers under one process
14.	Provided Sample Modules
14-a.	IMDB Parser
14-b.	Quotes mod with mysql/ini file system
14-c.	Simple http server
14-d.	Bash.org Parser
14-e.	News System/Rules Script
14-f.	Seen System (beta!)
14-g.	Request Ads System
15. 	Function Reference
16.	Special Thanks

=================================
1. Introduction and Release Notes
=================================
Please visit our new website: http://www.phpbots.org/ for support and user-submitted modules.

PHP-IRC is a totally php based irc bot meant to automate some applications related to irc.  I've personally programmed several modules which I currently use on irc.rizon.net.  This version is my attempt at bringing such a device to the public.  I hope you find it as useful as I do.  I have begun work on a module submission site which will be hosted at http://www.phpbots.org once finished.  I have already recieved modules from serveral enthusiastic people wishing to contribute in some way to the project.  Thank you for your time and effort.

It has always been my opinion that IRC scripting has long needed a simple programming interface that people with already existent skills could utilize.  There are two main reasons that PHP was selected as the primary language.  First off, I wanted to provide novice programmers a way to code their own functions and algorithms into a powerful platform without having to worry about such things as memory management and compilers.  Secondly, I wanted to do something that no one has done before--create a fully featured bot in php.   I don't know if I will ever complete the second objective, but I will keep on developing, as it will always be the main goal of this project.    

If I have left out information in this file, or there is something I said that confused you, please email me your concerns at manick@manekian.com, and I will try to clarify where you were confused.  This will also help me determine what I need to fix in this file for the next version.


=========================
1-a. What's new in 2.2.1!
=========================
As this is a service update and not a full release, the readme file included below has not been altered from the 2.2.0 version.  Version 2.2.1 brings several bug fixes (which are viewable in changes.txt).

Development on PHP-IRC is not halted, simply stalled, as I am approaching the culmination of my college career.  Thank you for your continued interest in PHP-IRC.

We have a new website!  Please view it here: http://www.phpbots.org/

=============
1-b. Features
=============
For a full list of current major features, please visit: http://www.phpbots.org/

===========
2. Included
===========

Core Files:
	==============
	Config Files
	--------------
	bot.conf
	function.conf
	typedefs.conf (added 2.2.0)
	modules/default/priv_mod.conf (added 2.2.0)
	modules/default/dcc_mod.conf (added 2.2.0)
	==============
	Readme Files
	--------------
	readme.txt
	modules/template.txt (added 2.1)
	gpl.txt (added 2.1)
	command_reference.txt (added 2.1.1)
	changes.txt (added 2.1.1)
	upgrade-2.1.1-2.2.0.txt  (added 2.2.0)
	install.txt  (added 2.2.0)
	==============
	Source code
	--------------
	bot.php
	chat.php (added 2.1)
	connection.php (added 2.2.0)
	dcc.php
	defines.php (added 2.0.1/2)
	error.php
	file.php (added 2.1)
	irc.php
	module.php (added 2.2.0)
	parser.php
	queue.php (added 2.2.0)
	remote.php (added 2.2.0)
	socket.php
	timers.php
	modules/default/dcc_mod.php (added 2.1)
	modules/default/priv_mod.php (added 2.1)

==========================================
3. Installation - A few things to consider
==========================================

I've been getting a lot of emails and private messages on rizon about how to run this bot via a web browser.  However, most webhosts do not compile php with socket support enabled, and for that matter, most web hosts don't have php5 as of yet.  Even so, in order to start the bot, you have to specify a config file on the command line.  Thus, running this bot from a web browser without modification is impossible.  I will not make this possible, so in other words, please stop asking.

Another thing that might deter people from using this script is the enourmous memory requirement by php to keep
a script loaded in memory.  PHP5 loads all of the program and everything it could possibly use into memory.  This can take up quite a bit of space!  I've had a bot use 8 megs of ram on the low end, and up to 15 megs on the high end.  If you don't mind dropping 15 megs of ram for this bot, then continue on.  Otherwise, don't whine to me that its a resource hog.  I believe the advantages far outweigh the disadvantages--how else could I have gotten this far?

In order to run this bot, you must have shell access to the box you will be running it on.  For instructions on how to install php5 for use with this bot, please see the next section, "Installing PHP 5".

A note about errno constants:
-----------------------------

In unix/windows/etc, when trying to write or read from a socket sometimes (a socket is the connection between your computer and another computer), errors occur.  In order to be able to determine what these errors are, all systems have something called 'errno' constants.  Unfortunatly, PHP5 does not know these constants, so we have to set them manually in order for this program to run.  In the defines.php file, I have set common values for windows, freebsd, linux, and unix.  If you are using one of these systems, an autodetection script should take care of detecting which system you are running the bot on, and set the appropriate values.  If this does not work, and you cannot connect to a server, you may need to set your OS type to either "windows", "freebsd", "linux", or "unix".  If you do not experience problems with this, then skip down to "Installing PHP 5".  Otherwise, keep reading.

I have run this successfully on Windows XP, Mandrake 9.0 Linux, and Debian.  If this program does not work with any of the os settings for OS in defines.php, you need to find the correct values for EAGAIN, EINPROGRESS, EALREADY, and EISCONN(these are those error constants that I was talking about earlier) for your system.

They are usually in /usr/include. You can find them by using a command sequence such as this:

-------------------------------------------------------------------------------
grep -r 'EAGAIN' /usr/include/*
grep -r 'EINPROGRESS' /usr/include/*
grep -r 'EALREADY' /usr/include/*
grep -r 'EISCONN' /usr/include/*
-------------------------------------------------------------------------------

After doing this, you'll get a line that looks like this:

-------------------------------------------------------------------------------
/usr/include/asm/errno.h:#define        EISCONN         106     /* Transport endpoint is already connected */
-------------------------------------------------------------------------------

The constant we are looking for is 106.  Now, after finding all those numbers, go into defines.php.  Set your OS type to "unknown"; then, scroll down a bit till you find this code:

-------------------------------------------------------------------------------
if ($OS == 'unknown')
{
	define('EAGAIN', 	0); 	/* Try again */
	define('EISCONN', 	0);	/* Transport endpoint is already connected */
	define('EALREADY',	0);	/* Operation already in progress */
	define('EINPROGRESS',	0);	/* Operation now in progress */
}
-------------------------------------------------------------------------------

Now, change the numbers, etc, with the corresponding values you got when you found the constants above.  Do this with all four constants, and then move on to the next section.

==================
3-a. Quick Install
==================

Running php-irc consists of 6 general steps (EXPERTS ONLY!):

1) Download and unpack, compile the php 5 5.0.2+ package with sockets and pcntl.
2) Edit the php.ini file following the steps in section 3-b (bottom half of section under "First, locate this line in the file:")
3) Unzip php irc
4) If the bot does not start up and connect to the server, you may have to set the OS parameter in the defines.php file.  Read the previous section.
5) Change server/port information in bot.conf.  You can run php-irc with the -c switch to get an md5 password hash to set dccadminpass with.
6) run php-irc with: /path/to/php/php.exe bot.php bot.conf from the php-irc directory, or edit the shebang line and change it to reflect your php location, (also use -c switch to specify php.ini location), and then chmod the bot.php file 755, and run with ./bot.php bot.conf

Of course, I recommend that you read the rest of this file, as it provides some very important information.

=====================
3-b. Installing PHP 5
=====================

Please read the following notices before continuing to install:

Notice 1:
---------
Do not use this bot with alpha/beta versions of php5.  I have successfully run it on rc2/rc3, however.

Notice 2:
---------
Do not use this bot with php 5.0.0 or 5.0.1.  These versions have a bug which makes it work incorrectly.  If you would like to find out more about this bug, view this page: http://bugs.php.net/bug.php?id=28892

Installation: (see below for installations on linux)

---------------------
Installing on Windows
---------------------
Download PHP5 from this location: 'http://www.php.net/downloads.php'.  Select the latest PHP 5 zip package under "Windows Binaries".  Unzip this to a folder, c:\php, for instance.

Then, copy the included 'php.ini-dist' file in the c:\php directory to 'c:\windows\php.ini'.  Open this file with notepad or your favorite text editor. (Usually, you can right click the file and hit 'open with').

First, locate this line in the file:
---------------------------------------------------------------------------
; This directive tells PHP whether to declare the argv&argc variables (that
---------------------------------------------------------------------------

A few lines below this, you'll see this directive:
--------------------------------
register_argc_argv
--------------------------------
Make sure that this line says:
--------------------------------
register_argc_argv = On
--------------------------------

Now, locate this line:
--------------------------------
extension_dir = "./"
--------------------------------
Change this to:
--------------------------------
extension_dir = "./ext"
--------------------------------

If you are going to use a mysql database, find:
------------------------
;extension=php_mysql.dll
------------------------
And change to:
------------------------
extension=php_mysql.dll
------------------------

Then find:
--------------------------------
;extension=php_sockets.dll
--------------------------------
And change to:
--------------------------------
extension=php_sockets.dll
--------------------------------

You are now set to run PHP-IRC.  See section "Configuration" for information related to configuring the bot.

-------------------
Installing on Linux
-------------------
Download php from php website, here: 'http://www.php.net/downloads.php'.  Select the latest php5 tar.gz package under "Complete Source Code".

Extract this file.  To do this, copy the file to your home directory.  Then, extract it using this command:

tar -zxf php-<your version>.tar.gz

Replace <your version> with your version of PHP.  Make sure it is 5.0.2 or greater!
Now, run this command:

cd php-<your version>/

Also make sure to replace <your version> with your php5 version.
Now, we have a few things to consider here.  Do you want to run a mysql database or not?  If you do, then use this command:

./configure --enable-sockets --enable-pcntl --with-mysql

However, if you do not have mysql installed on your system, do this:

./configure --enable-sockets --enable-pcntl

If for any reason this fails, please read below.  If it does not, skip these next few paragraphs.

Usually the configure script fails when it cannot find a module or package it needs to compile.  Sometimes I have had it fail with the 'XML LIB' package, saying it could not be found.  In order to rectify this, use this command:

./configure --help

And then search through the output, looking for the package that failed.  See if there is a --disable-<package> option that you can add to the ./configure command above, which will make it skip that package.  Then, run the configure command again, until you remove all packages that do not work with your system.

Continuing on
-------------

Now, after you see the "Thanks for using PHP" message, run the following command:

make

After this is completed, you are all done installing php!  You will need to copy the binary to a usable and NON WEB ACCESSABLE directory, by using this command:

cp sapi/cli/php /<my directory>/

Where <my directory> is the directory you wish to copy it to.  When I say NON WEB ACCESSABLE, I mean that you could not access this file directly from the web.

You are now set to run PHP-IRC.  See section "Configuration" for information related to configuring the bot.

=====================================================
3-c. Using alternate PHP.INI (PHP Configuration) file
=====================================================

You may be running another version of php on the server that you run this on specifically for apache, or some other purpose.  In this case, you already have a php.ini file in /etc.  You can use an alternate php.ini file by using this syntax when running php:

/path/to/php.exe -c /dir/of/php.ini/ bot.php bot.conf

If on linux, you can also chmod bot.php 755, and then edit the #! line at the top of the file to look like this:

#!/path/to/php -c /dir/of/php.ini/

And then you can run php-irc like this:

./bot.php bot.conf

=======================
3-d. Installing PHP-IRC
=======================
If you are reading this file, you have already downloaded and unpacked the source package from sourceforge.net.  Good, the hard part is out of the way.  Continue on to the next section.


================
4. Configuration
================

PHP5
----
Please see above, our guide to "installing php 5" for configuration options to php.ini.

defines.php
-----------
The main things you need to look at in here is DEBUG mode and OS. DEBUG 1 will make it so everything that happens to the bot is printed in the main window.  DEBUG 0 will instead print it to a logfile specified in bot.conf. Make sure that you set your OS to either linux, unix, windows or bsd, otherwise your bot won't run (if autodetection fails).

The lines you need to change are:

define('DEBUG', 1);

and:

define('OS', 'windows');

You might also want to change the name of the PID file that is written when you run the bot in background mode.  You can edit that filename with this line:

define('PID', "bot.pid");

bot.conf
--------
Edit this file and change all the options to your liking.  Of special interest should be natip, which allows you to work from behind a nat.  Also, as of 2.1, pay special attention to the 'upload' and 'uploaddir' parameters.  You can accept file transfers with those.  Please also look at dccadminpass.  You will need to set this in order to use dcc chat or private message administration of your bot.  You will need to run 'bot.php -c password' to generate an md5 password hash of 'password'.

function.conf
-------------
This file is a little more complicated.  php-irc will respond to various text typed in a channel (triggers).  You can configure those triggers in this file.  Please see the section titled "Modules and User-defined functions" for help with this file.

At this point, you should be able to start up your bot--although it won't do much.  See the section "Modules and User-defined functions" for information about adding functionality.


==========
5. Running
==========
Running this bot is rather simple.

Windows:
--------

Run the bot with the following command:

c:\php\php.exe bot.php bot.conf

Make sure that this is the correct path that php is in.  You may need to change 'c:\php' to the directory where your php5's php.exe resides.  Make sure, also, that you are located in your php-irc main directory when you start the bot. i.e.,

cd <my php-irc directory>

I have a little trick.  I make a windows shortcut (.lnk file) to php.exe, and put it in my php-irc directory.  Then I edit the shortcut and make sure that my working directory is my php-irc directory.  Then, I can run the bot as such:

php.lnk bot.php bot.conf

Linux:
------
The basic syntax is:

/path/to/php5/php bot.php bot.conf

If you followed my php5 installation instructions, /path/to/php5/ would be the <my directory> that we talked about earlier.

Linux Alternative:

You can edit the #! line in bot.php to reflect your php cgi binary location, and 'chmod 755 bot.php'.  This will make 'bot.php' executable.  You can then run it as such:

./bot.php bot.conf

You can also run several different configurations under the same process.  Simply make another bot.conf file and then run it like this:

./bot.php bot.conf bot2.conf

If you are on linux/unix, you can also run in the background by using the -b switch:
./bot.php -b bot.conf

changes.txt
+---------------------------------------------------------------------------
|   PHP-IRC v2.2.3 Service Release
|   ========================================================
|   by Manick
|   (c) 2001-2005 by http://www.phpbots.org/
|   Contact: manick@manekian.com
|   irc: #manekian@irc.rizon.net
|   ========================================
|   Maintained by g2x3k
|   2011-2015 https://github.com/g2x3k/php-irc
|   contant: g2x3k@layer13.net
|   irc: #root @ irc.layer13.net:+7000
+---------------------------------------------------------------------------
|   > remote class module
|   > Module written by Manick
|   > Module Version Number: 2.2.0
+---------------------------------------------------------------------------
|   > This program is free software; you can redistribute it and/or
|   > modify it under the terms of the GNU General Public License
|   > as published by the Free Software Foundation; either version 2
|   > of the License, or (at your option) any later version.
|   >
|   > This program is distributed in the hope that it will be useful,
|   > but WITHOUT ANY WARRANTY; without even the implied warranty of
|   > MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
|   > GNU General Public License for more details.
|   >
|   > You should have received a copy of the GNU General Public License
|   > along with this program; if not, write to the Free Software
|   > Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+---------------------------------------------------------------------------
|   Changes
|   =======-------
|   > If you wish to suggest or submit an update/change to the source
|   > code, post a pull request or issue on github and i will into it
|   > https://github.com/g2x3k/php-irc
|   >                                            maintained by g2x3k
+---------------------------------------------------------------------------


+ taken over by g2x3k see https://github.com/g2x3k/php-irc/commits/master for changes

- Added: New website: http://www.phpbots.org/
- Added: "$this->getModule($modName)" for modules, so that they can access other modules directly
- Added: Ability to spawn new bot from admin interface
- Added: Ability to bind to IP when connecting to server
- Added: Ability to connect to passworded servers
- Added: 'raw' module option, where all messages of any type are passed to the module specified
- Added: 'action' module option, to capture the '/me' tag.

- Changed: Added news module and added "!news del high" and "!news del low" commands
- Changed: Fixed bug in chat.php where "true" was mistyped as "ture" in handshaketimeout function
- Changed: Fixed bug in irc::myStrToLower and irc::myStrToUpper
- Changed: Fixed bug where 004 wasn't be interpreted correctly as 'connect' in function files
- Changed: Fixed bug where bot trys to identify when asked to by nickserv and password field commented out in config file.
- Changed: Fixed Major bug where onRead() was being done before onAccept() was called on a new created listen socket.


bot.php
<?php
/*
+---------------------------------------------------------------------------
|   PHP-IRC v2.2.3 Service Release
|   ========================================================
|   by Manick
|   (c) 2001-2005 by http://www.phpbots.org/
|   Contact: manick@manekian.com
|   irc: #manekian@irc.rizon.net
|   ========================================
|   Maintained by g2x3k
|   2011-2015 https://github.com/g2x3k/php-irc
|   contant: g2x3k@layer13.net
|   irc: #root @ irc.layer13.net:+7000
+---------------------------------------------------------------------------
|   > Main module
|   > Module written by Manick
|   > Module Version Number: 2.2.0
+---------------------------------------------------------------------------
|   > This program is free software; you can redistribute it and/or
|   > modify it under the terms of the GNU General Public License
|   > as published by the Free Software Foundation; either version 2
|   > of the License, or (at your option) any later version.
|   >
|   > This program is distributed in the hope that it will be useful,
|   > but WITHOUT ANY WARRANTY; without even the implied warranty of
|   > MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
|   > GNU General Public License for more details.
|   >
|   > You should have received a copy of the GNU General Public License
|   > along with this program; if not, write to the Free Software
|   > Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+---------------------------------------------------------------------------
|   Changes
|   =======-------
|   > If you wish to suggest or submit an update/change to the source
|   > code, post a pull request or issue on github and i will into it
|   > https://github.com/g2x3k/php-irc
|   >                                            maintained by g2x3k
+---------------------------------------------------------------------------
*/

error_reporting(E_ALL);
set_time_limit(0);

require('./defines.php');
require('./queue.php');
require('./module.php');
require('./irc.php');
require('./socket.php');
require('./timers.php');
require('./dcc.php');
require('./chat.php');
require('./file.php');
require('./parser.php');
require('./databases/ini.php');
require('./error.php');
require('./connection.php');
require('./remote.php');

// include other files
if ($handle = opendir('./includes/')) {
    while (false !== ($entry = readdir($handle))) {
        if ($entry != '.' && $entry != '..') {
            if (preg_match('/.php$/i', $entry))
                include('./includes/' . $entry);
        }
    }
    closedir($handle);
}

final class bot
{

    /* Global socket class used by all bots */
    private $socketClass;

    /* Global process Queue used by all bots, timers, dcc classes */
    private $procQueue;

    /* Whether we are running in background mode or not.  (not sure if this is used anymore */
    private $background = 0;

    //contain all the bots
    private $bots = array();

    // save the only one instance of bot (singleton)
    private static $_instance;

    public static function getInstance()
    {
        if (!isset (self:: $_instance)) {
            self:: $_instance = new bot();
        }
        return self:: $_instance;
    }


    //Main Method
    private function __construct()
    {

        $this->socketClass = new socket();
        $this->procQueue = new processQueue();

        $this->socketClass->setProcQueue($this->procQueue);

        $this->readConfig();
    }

    public function launch()
    {

        foreach ($this->bots as $bot) {
            $this->createBot($bot);
        }

        try {

            /* Main program loop */

            while (1) {
                //Get data from sockets, and trigger procQueue's for new data to be read
                $this->socketClass->handle();

                //The bots main process loop.  Run everything we need to!
                $timeout = $this->procQueue->handle();

                //Okay, set the socketclass timeout based on the next applicable process
                if ($timeout !== true) {
                    $this->socketClass->setTimeout($timeout);
                }

                //echo $this->procQueue->getNumQueued() . "\n";
                //$this->procQueue->displayQueue();

                //Aight, if we don't have any sockets open/in use, and
                //we have no processes in the process queue, then there are
                //obviously no bots running, so just exit!
                if ($this->socketClass->getNumSockets() == 0 && $this->procQueue->getNumQueued() == 0) {
                    break;
                }
            }

        } catch (Exception $e) {
            $this->ircClass->log($e->_toString());
        }

    }


    public static function addBot($filename)
    {
        $bot = bot::getInstance();
        $config = bot::parseConfig($filename);
        if ($config == false) {
            return false;
        }

        $newBot = new botClass();
        $newBot->config = $config;
        $newBot->configFilename = $filename;

        $bot->bots[] = $newBot;
        $bot->createBot($newBot);
        return true;
    }


    private function createBot($bot)
    {
        $this->connectToDatabase($bot);

        $bot->socketClass = $this->socketClass;
        $bot->timerClass = new timers();
        $bot->parserClass = new parser();

        $bot->dccClass = new dcc();
        $bot->ircClass = new irc();

        $bot->ircClass->setConfig($bot->config, $bot->configFilename);
        $bot->ircClass->setSocketClass($this->socketClass);
        $bot->ircClass->setParserClass($bot->parserClass);
        $bot->ircClass->setDccClass($bot->dccClass);
        $bot->ircClass->setTimerClass($bot->timerClass);
        $bot->ircClass->setProcQueue($this->procQueue);

        $bot->dccClass->setSocketClass($this->socketClass);
        $bot->dccClass->setTimerClass($bot->timerClass);
        $bot->dccClass->setParserClass($bot->parserClass);
        $bot->dccClass->setProcQueue($this->procQueue);

        $bot->parserClass->setTimerClass($bot->timerClass);
        $bot->parserClass->setSocketClass($this->socketClass);
        $bot->parserClass->setDatabase($bot->db);

        $bot->timerClass->setIrcClass($bot->ircClass);
        $bot->timerClass->setSocketClass($this->socketClass);
        $bot->timerClass->setProcQueue($this->procQueue);

        $bot->parserClass->init();

        //Okay, this function adds the connect timer and starts up this bot class.
        $bot->ircClass->init();

        bot::createChannelArray($bot->ircClass);

    }


    private function readConfig()
    {
        global $argc, $argv;

        if ($argc < 2) {
            echo "You must specify a config file.\n";
            die();
        }

        $isPasswordEncrypt = false;

        array_shift($argv);

        foreach ($argv AS $filename) {
            if ($filename == "") {
                continue;
            }

            if ($isPasswordEncrypt == true) {
                die("Encrypted Password: " . md5($filename) . "\nReplace this as 'dccadminpass' in bot.conf!");
            }

            if ($filename == "-c") {
                $isPasswordEncrypt = true;
                continue;
            }

            if ($filename == "-b" && $this->background != 1) {
                $this->background == 1;
                $this->doBackground();
                continue;
            }

            $config = bot::parseConfig($filename);

            if ($config == false) {
                echo "Could not spawn bot $filename";
                die();
            }

            $bot = new botClass();
            $bot->config = $config;
            $bot->configFilename = $filename;

            $this->bots[] = $bot;
        }

        if ($isPasswordEncrypt == true) {
            die("No password submitted on command line! Syntax: bot.php -c <new admin password>\n");
        }

    }


    private function connectToDatabase($bot)
    {
        if (isset($bot->config['usedatabase'])) {
            if (!file_exists("./databases/" . $bot->config['usedatabase'] . ".php")) {
                die("Couldn't find the database file! Make sure it exists!");
            }

            require_once("./databases/" . $bot->config['usedatabase'] . ".php");

            $dbType = $bot->config['usedatabase'];

            if (!isset($bot->config['dbhost']))
                $bot->config['dbhost'] = "localhost";
            if (!isset($bot->config['dbuser']))
                $bot->config['dbuser'] = "root";
            if (!isset($bot->config['dbpass']))
                $bot->config['dbpass'] = "";
            if (!isset($bot->config['db']))
                $bot->config['db'] = "test";
            if (!isset($bot->config['dbprefix']))
                $bot->config['dbprefix'] = "";
            if (!isset($bot->config['dbport'])) {
                $bot->db = new $dbType($bot->config['dbhost'],
                    $bot->config['db'],
                    $bot->config['dbuser'],
                    $bot->config['dbpass'],
                    $bot->config['dbprefix']);
            } else {
                $bot->db = new $dbType($bot->config['dbhost'],
                    $bot->config['db'],
                    $bot->config['dbuser'],
                    $bot->config['dbpass'],
                    $bot->config['dbprefix'],
                    $bot->config['dbport']);
            }

            if (!$bot->db->isConnected()) {
                die("Couldn't connect to database...");
            }

        }
    }


    public static function createChannelArray($ircClass)
    {
        $channels = $ircClass->getClientConf('channel');

        if ($channels != "") {
            if (!is_array($channels)) {
                $channels = array($channels);
            }

            foreach ($channels AS $channel) {
                $chan = $channel;
                $key = "";

                if (strpos($channel, chr(32)) !== false) {
                    $channelVars = explode(chr(32), $channel);
                    $chan = $channelVars[0];
                    $key = $channelVars[1];
                }

                $ircClass->maintainChannel($chan, $key);
            }
        }
    }

    public static function parseConfig($filename)
    {
        $configFPtr = @fopen($filename, "rt");

        if ($configFPtr == null) {
            echo "Could not find config file '" . $filename . "'\n";
            return false;
        }

        $configRaw = "";

        try {

            while (!feof($configFPtr)) {
                $configRaw .= fgets($configFPtr, 1024);
            }

            fclose($configFPtr);

        } catch (Exception $e) {
            //	echo "A fatal IO Exception occured.";
            return false;
        }

        $config = array();

        $configRaw = str_replace("\r", "", $configRaw);

        $confLines = explode("\n", $configRaw);

        foreach ($confLines AS $line) {
            $line = trim($line);

            if ($line == "" || substr($line, 0, 1) == ";") {
                continue;
            }

            $offsetA = strpos($line, chr(32));

            if ($offsetA != false) {
                $confVar = substr($line, 0, $offsetA);
                $confParams = substr($line, $offsetA + 1);
            } else {
                $confVar = $line;
                $confParams = "";
            }

            if (isset($config[$confVar])) {
                if (!is_array($config[$confVar])) {
                    $prevParam = $config[$confVar];
                    $config[$confVar] = array();
                    $config[$confVar][] = $prevParam;
                }

                $config[$confVar][] = $confParams;

            } else {
                $config[$confVar] = $confParams;
            }
        }

        return $config;
    }


    private function doBackground()
    {
        $pid = pcntl_fork();
        if ($pid == -1) {
            die("Error: could not fork\n");
        } else if ($pid) {
            if (PID != "") {
                $file = fopen(PID, "w+");
                fwrite($file, $pid);
                fclose($file);
            }

            exit(); // Parent
        }

        if (!posix_setsid()) {
            die("Error: Could not detach from terminal\n");
        } else {
            echo "PHP|IRC is now running in the background\n";
        }

        fclose(STDOUT);

    }

}

$ircBot = bot::getInstance();
$ircBot->launch();

?>
